### PR TITLE
refactor(cron): remove saffron dependency

### PR DESCRIPTION
**Pull request by nathanwhitbot**
## Summary
- Replace the `saffron` dependency in `deno_cron` with a purpose-built internal cron parser and scheduler.
- Preserve saffron-compatible parsing and matching behavior, including numeric weekday mapping, plain and stepped reverse ranges, `L`/`LW`/`L-NW`, nearest-weekday, and `n#k` cases.
- Remove `saffron` and its old `nom v5` dependency chain from the workspace dependency graph.
- Add table-driven regression coverage for accepted schedules, rejected malformed schedules, saffron-compatible edge cases, panic-safety around invalid derived dates, and wraparound range behavior.

## Validation
- `cargo test -p deno_cron`
- `cargo test -p specs_tests --test specs cron` passed for the non-socket cron specs; socket cron specs failed due to missing `tests/util/std/path/mod.ts` fixtures in this checkout before cron logic ran.
- `cargo test -p unit_tests --test unit cron_test` was blocked by missing `tests/util/std/...` fixtures before cron logic ran.
- `cargo tree -i saffron` confirmed absent.
- `cargo tree -i nom@5.1.3` confirmed absent.
- Debug binary size check against a scratch `HEAD` baseline showed no increase; the current build was smaller by 2,499,648 bytes.

### CODE REVIEW

File: ext/cron/cron.rs [hunk 1]
```diff
@@ -0,0 +1,678 @@
...
+    "AUG" => Ok(8),
+    "SEP" => Ok(9),
+    "OCT" => Ok(10),
+    "NOV" => Ok(11),
+    "DEC" => Ok(12),
+    _ => parse_number(input, 1, 12),
+  }
+}
+
+fn parse_day_of_week(input: &str) -> Result<u32, ParseError> {
```
Comment (bartlomieju) [inline ext/cron/cron.rs:410]: The numeric weekday mapping is offset relative to the named one: numeric `1` -> Sunday, `2` -> Monday, ..., while `MON` -> Monday. So `1` and `MON` resolve to *different* days. This looks intentional (saffron's quirk) and the frozen tests assert it, but it's surprising enough to deserve a one-line comment here. Can you confirm this matches actual saffron output?

File: ext/cron/cron.rs [hunk 2]
```diff
@@ -0,0 +1,678 @@
...
+    "FRI" => Ok(5),
+    "SAT" => Ok(6),
+    _ => match parse_number(input, 1, 7)? {
+      7 => Ok(6),
+      day => Ok(day - 1),
+    },
+  }
+}
+
+fn range_bits(start: u32, end: u32, field: Field) -> u64 {
```
Comment (bartlomieju) [inline ext/cron/cron.rs:428]: Plain reverse ranges subtract 1 here (`bit_range(start - 1, ...)`), so `59-0` matches minutes 58, 59, 0 and `SAT-SUN` matches Fri/Sat/Sun -- but `stepped_bits` just below does *not* apply this off-by-one (`20-4/2` starts at 20). This asymmetry is exactly the kind of thing a future reader will try to "fix" and silently break compat. A short comment on both functions explaining it's a deliberate saffron-compat quirk would help.

File: ext/cron/cron.rs [hunk 3]
```diff
@@ -0,0 +1,678 @@
...
+  }
+}
+
+impl Schedule {
+  pub fn next_after(&self, date: DateTime<Utc>) -> Option<DateTime<Utc>> {
+    let mut next = date
+      .with_nanosecond(0)?
+      .with_second(0)?
+      .checked_add_signed(Duration::minutes(1))?;
+    let end = next.checked_add_signed(Duration::days(366 * 5))?;
```
Comment (bartlomieju) [inline ext/cron/cron.rs]: nit: the search window is `366 * 5` days (~5 years). A Feb-29 schedule crossing a non-leap century boundary (2096 -> 2104 is an 8-year gap) would wrongly return `None`. Won't matter before ~2096, but worth a comment noting the bound, or matching whatever window saffron used.

**Review by bartlomieju — COMMENTED** [review]
Clean, idiomatic, panic-safe replacement for `saffron`, and it nicely drops both `saffron` and the orphaned `nom 5.1.3` from the workspace dep graph (the `Cargo.lock` `nom 7.1.3` -> `nom` renames are what cargo generates once only one version remains). It satisfies the existing `tests/unit/cron_test.ts` expectations, and the panic-safety test (`L-30W APR`, `31 FEB`) is a nice touch.

The risk in a behavior-preserving refactor like this lives entirely in the subtle saffron quirks. The PR faithfully replicates several non-obvious ones (offset numeric weekday mapping, the `start-1` quirk on plain reverse ranges, the *absence* of that quirk on stepped reverse ranges, and OR semantics when both day-of-month and day-of-week are restricted). The whole correctness argument rests on whether the "frozen" expected values were captured from actual saffron output rather than hand-assumed.

My main ask: please confirm the frozen cases were generated by running real saffron (e.g. a one-off comparison harness), so the tests encode saffron's behavior and not our assumptions. See inline comments for two spots where a short clarifying comment would protect the quirk from a well-meaning future "cleanup".

Not blocking, just flagging: CI hasn't reported any checks yet, and the second commit message ("The cron crate tests pass locally. I'm doing a final diff/status read") reads like a stray note. Both wash out on squash-merge, but CI should be green first.

**Comment by nathanwhit** [comment]
Ran a differential fuzz of this `Schedule` impl against the real `saffron 0.1.0` crate — 300k randomly-generated expressions (biased toward edge syntax: wraparound ranges, steps, `L`/`LW`/`L-NW`/`NW`/`nL`/`n#k`, out-of-range values, names), comparing both parse accept/reject and `next_after` over 12 base timestamps each.

**The compat story holds up well:**

- **Parse parity is perfect** — 0 disagreements over 300k exprs (184,961 both-accept / 115,039 both-reject, zero one-sided).
- **`next_after` is identical on every realistic schedule** — plain ranges (incl. reverse), lists, names, `L`/`LW`/`L-NW`/`NW`/`nL`/`n#k` all match saffron exactly.
- The new impl is **strictly safer**: it handles 530+ inputs where saffron *panics* (`"Month should be valid"` on reverse stepped month ranges like `0 0 1 10-2/3 *`).

Of the ~0.7% `next_after` divergences, ~99% are wrapping stepped ranges (`a-b/n` with a>b) where saffron itself panics or emits inconsistent sets — i.e. this PR is the improvement, not a regression.

**One real regression worth fixing before merge — the search horizon.**

`next_after` gives up after `366 * 5` days, so it returns `None` for valid-but-rare schedules whose next firing is further out:

```
0 0 * 2 FRI#5   (5th Friday of February)
  saffron : 2036-02-29   ✓
  schedule: <none>       ✗   (next occurrence is 16 years out)
```

Same failure mode hits `0 0 29 2 *` evaluated in the years before a skipped leap year (2100/2200/2300 → 8-year gap). This matters in `compute_next_deadline`, which turns `None` into `CronError::InvalidCron` — so a schedule that worked under saffron would now be rejected.

Suggest widening or removing the `366 * 5` cap (Feb-29 needs ~8y, `<wd>#5`-in-Feb needs ~28y; or just loop to a far cap like +100y), plus a regression test:

```rust
assert!(
  "0 0 * 2 FRI#5".parse::<Schedule>().unwrap()
    .next_after(dt(2020, 1, 1, 0, 0)).is_some()
);
```

**Comment by nathanwhit** [comment]
Re-ran the saffron differential fuzz against the updated branch (`6519cde`) — **the horizon regression is fixed.** Verified against real saffron:

```
0 0 * 2 FRI#5      -> 2036-02-29   ✓ (was <none>)
0 0 29 2 * @2096   -> 2104-02-29   ✓ (century leap gap)
```

and the added regression tests cover exactly these. Parse parity is still perfect (0 one-sided accepts over 300k exprs), and the 530+ saffron-panic inputs remain handled safely.

One small residual difference I pinned down while characterizing the leftovers — **not a blocker, just flagging so the behavior is deliberate.** `W` on a day the month doesn't have diverges:

```
0 0 31W 4 *   (nearest weekday to Apr 31 — April has 30 days)
  saffron : fires on the 30th
  schedule: never matches (None)
```

`closest_weekday()` returns `None` when `day > days_in_month`, so `31W` in any 30-day month (and `30W`/`31W` in Feb) never fires, whereas saffron clamps to the last day. It's genuinely ambiguous which is correct, and saffron is itself inconsistent here (`0 0 31W * *` skips April, but with the month pinned to April it fires the 30th) — so I don't think the PR is wrong. Worth a deliberate call rather than a fix.

Otherwise LGTM — every realistic schedule matches saffron, and the remaining divergences are all the wrapping-range cases where saffron panics or emits inconsistent output.
